core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java

```diff
@@ -94,39 +102,65 @@ public boolean recordCursorSupportsRandomAccess() {
     }
 
     public void recalculateIncludedValues(TableReader tableReader) {
-        excludedValues.clear();
-        for (int i = 0, n = keyExcludedValueFunctions.size(); i < n; i++) {
-            excludedValues.add(Chars.toString(keyExcludedValueFunctions.getQuick(i).getStr(null)));
+        excludedKeys.clear();
+        if (dynamicExcludedKeys) {
+            // In case of bind variable excluded values that may change between
+            // query executions the sets have to be subtracted from scratch.
+            includedKeys.clear();
+            cursorFactories.clear();
         }
+
         final SymbolMapReaderImpl symbolMapReader = (SymbolMapReaderImpl) tableReader.getSymbolMapReader(columnIndex);
-        for (int i = 0, n = symbolMapReader.getSymbolCount(); i < n; i++) {
-            final CharSequence symbol = symbolMapReader.valueOf(i);
-            if (excludedValues.indexOf(symbol) < 0 && includedValues.indexOf(symbol) < 0) {
-                final RowCursorFactory rowCursorFactory;
-                int symbolKey = symbolMapReader.keyOf(symbol);
-                if (filter == null) {
-                    rowCursorFactory = new SymbolIndexRowCursorFactory(
-                            columnIndex,
-                            symbolKey,
-                            cursorFactories.size() == 0,
-                            indexDirection,
-                            null
-                    );
-                } else {
-                    rowCursorFactory = new SymbolIndexFilteredRowCursorFactory(
-                            columnIndex,
-                            symbolKey,
-                            filter,
-                            cursorFactories.size() == 0,
-                            indexDirection,
-                            columnIndexes,
-                            null
-                    );
-                }
-                includedValues.add(Chars.toString(symbol));
-                cursorFactories.add(rowCursorFactory);
+
+        boolean excludeNull = false;
+        // Generate excluded key set.
+        for (int i = 0, n = keyExcludedValueFunctions.size(); i < n; i++) {
+            final CharSequence value = keyExcludedValueFunctions.getQuick(i).getStr(null);
+            if (value != null) {
```

Comment:
I don't think this `if` is necessary. `keyOf` can work with nulls

Reply:
Addressed in b1075cc62e628100e930c525cdc36b94294761de